### DISCUSSION TITLE
client.get_channel() fails intermittently when user ID is supplied

### DISCUSSION THREAD

**Discussion by yojc** [Rapptz/discord.py#7152] [Q&A] [answered]
Hi!

One of the functions my bot has are reminders - a user can set a message to send from the bot at specified time. The message is sent to the channel, where the user message with the bot command has been sent (or DM, if the reminder was set via DM with bot).

This works well if the target message is sent to a channel on any guild. However, when trying to send a DM via bot, it only works sometimes. What's strange, I'm pretty sure it used to work reliably all the time a few months ago, but it doesn't anymore (?). I might have overlooked any relevant commits to the repo or closed issued in the bug tracker, but didn't find any.

I'm trying to get the destination channel via `client.get_channel()`, but when passing an user ID as a parameter it can sometimes return `None`, even though the ID is 1000000% correct. `client.fetch_user()` returns `NotFound`, `client.get_user()` returns `None`. Server Members Intent is turned on; events such as `on_member_join` are working correctly.

From I could gather, I think that `client.get_channel()` (when supplying user ID) and  `client.fetch_user()` / `client.get_user()` fail, when the target user haven't sent any DMs to the bot since the bot has been started. In other words, when I restart the bot, it won't be able to sent any DMs to the users that haven't messaged him privately since the restart. To be clear, I'm not 100% sure if it's always the case, but it seems very likely.

I'm at a loss. Is there something I need to set up before sending a DM to a user? What am I doing wrong? It's especially confusing since it used to work.

Here's the code if it's at all important. Sorry in advance, I know it's not something to be proud of. :P
https://github.com/yojc/BillyMays/blob/main/billy_c_sopel_remind.py#L82

**Comment by nihaals** [comment, marked as answer]
The intended way of getting the DM between the client and a user is `client.get_user(USER_ID).dm_channel`. A DM's channel ID is not the same as one of the user's IDs.

The [docs for `User.dm_channel`](https://discordpy.readthedocs.io/en/latest/api.html#discord.User.dm_channel) explains its return type and what to do if it returns `None`.

**Reply by AbstractUmbra** [reply]
Further to the above, `User.dm_channel` may be `NoneType` if one has not been opened or has been removed from the bot's internal cache due to inactivity.

You can use [`User.create_dm()`](https://discordpy.readthedocs.io/en/latest/api.html?highlight=create_dm#discord.User.create_dm) to create one if it does not exist.

Standard restrictions apply, such as the requirement to share at least one guild with the user in question, etc.

**Reply by yojc** [reply]
Okay, now that I've read your replies and taken a more thorough look at my code, I have narrowed down the bug, which was trivial - as always.

The bug description I wrote turned out to be incorrect. Actually, I haven't been calling `get_channel()` with the user ID as parameter - I have always been using the DM channel ID. It's only now that I realised these aren't the same... But anyway - now that I know that, it's no wonder that `fetch_user()` and `get_user()` failed - I have been using the DM ID there as well.

With that out of the way, the fix was pretty simple, and now the message sending algorithm is as follows:
1. Try to get the destination channel via `client.get_channel(DM_ID)`
2. If it returns `None`, try finding the user via `client.get_user(USER_ID)`
3. If the `dm_channel` is `None`, try `create_dm()`

From my brief testing, it seems like this has solved the issue. I'll need some more time to be completely sure, but it looks promising.

Also, I wrote that my prior code used to work some time ago, but now I have remembered that I didn't mention something important (sorry!). A few months ago I've finally upgraded from `Python 3.4` + `discord.py 0.16.12` to `Python 3.7.3` + `discord.py 1.6.0`. It's entirely possible that `0.16.12` handled things a little bit differently, and the refactor was not done properly. It might have never worked correctly after the upgrade for all that I know.

Anyway - thanks a lot 👍
